### PR TITLE
[GRPO] Allow the use of the vllm logprobs, rather than recomputing them

### PR DESCRIPTION
**Pull request by edbeeching**
This PR exposes the logprobs of token IDs that were generated from the vllm client. 
The default will be false, as they are likely different to those produced by the policy.

### CODE REVIEW

**Review by lewtun — CHANGES_REQUESTED** [review]
Overall looks good with a potential typo in the padding of the log probs. It would be quite interesting to know if there is any noticeable difference between using the vLLM logprobs vs native ones when `beta > 0` (not for this PR, but more generally on some simple baseline)

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -81,6 +81,8 @@ class GRPOConfig(TrainingArguments):
         use_vllm (`bool`, *optional*, defaults to `False`):
             Whether to use vLLM for generating completions. If set to `True`, ensure that a GPU is kept unused for
             training, as vLLM will require one for generation. vLLM must be installed (`pip install vllm`).
+        use_vllm_logprobs (`bool`, *optional*, defaults to `False`):
+            Whether to use vLLM's logprobs for the `"old_logprobs"` in the GRPO loss. Requires `use_vllm=True`.
```
[the remark this answers is not in the document]
Comment (edbeeching) [reply trl/trainer/grpo_config.py:85]: Thanks. By the way, I think there may be some confusion, this PR is not related to the ref model at all. Previously the `old_logprobs` were (re)calculated from the model we are optimizing in `torch.no_grad` mode, using the tokens generated from the vllm instance. They are used as part of GRPO's clipping loss to constrain the model's updates to be within beta of the old model. 

Now we use the logprobs from the vllm instance, I believe you are thinking of the KL penalty, which still uses the log_probs from the refmodel and the model we are optimizing. Not the vllm logprobs.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -721,18 +721,23 @@ def _generate_and_score_completions(
                     )
             else:
                 completion_ids = [None] * len(all_prompts_text)
+                vllm_log_probs = [None] * len(all_prompts_text)
             # Broadcast the completions from the main process to all processes, ensuring each process receives its
             # corresponding slice.
             completion_ids = broadcast_object_list(completion_ids, from_process=0)
+            vllm_log_probs = broadcast_object_list(vllm_log_probs, from_process=0)
```
[the remark this answers is not in the document]
Comment (edbeeching) [reply trl/trainer/grpo_trainer.py:728]: Most likely marginal as the GPUs all already in sync due to the gather of `completion_ids`.

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -721,18 +721,23 @@ def _generate_and_score_completions(
...
                 (self.accelerator.process_index + 1) * len(prompts),
             )
             completion_ids = completion_ids[process_slice]
+            vllm_log_probs = vllm_log_probs[process_slice]
 
             # Pad the completions, and concatenate them with the prompts
             completion_ids = [torch.tensor(ids, device=device) for ids in completion_ids]
             completion_ids = pad(completion_ids, padding_value=self.processing_class.pad_token_id)
+            vllm_log_probs = [torch.tensor(logp, device=device) for logp in vllm_log_probs]
+            vllm_log_probs = pad(vllm_log_probs, padding_value=self.processing_class.pad_token_id)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:740]:
```suggestion
            vllm_log_probs = pad(vllm_log_probs, padding_value=0.0)
```

**Review by qgallouedec — APPROVED** [review]
👌

File: trl/trainer/grpo_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lewtun) [reply trl/trainer/grpo_config.py:85]: Ah yeah you're right - I've amended my suggestion to align with that. (I think it's not a good idea to reference internal variables in docstrings as the user cannot see them unless they drill into the code)

**Review by lewtun — APPROVED** [review]
Thanks for the clarification about how the logprobs are used in the loss. LGTM with the minor tweak to the docstring

File: trl/scripts/vllm_serve.py [hunk 4]
```diff
@@ -321,6 +323,7 @@ async def generate(request: GenerateRequest):
         Example response:
         ```json
         {"completion_ids": [[101, 102, 103], [201, 202, 203]]}
+        {"log_probs": [[1.1, 1.2, 1.3], [2.1, 2.2, 2.3]]}
```
Comment (qgallouedec) [inline trl/scripts/vllm_serve.py:326]:
```suggestion
        {"completion_ids": [[101, 102, 103], [201, 202, 203]],
         "log_probs": [[1.1, 1.2, 1.3], [2.1, 2.2, 2.3]]}
```

**Comment by edbeeching** [comment]
I benchmarked this, and the performance is comparable but speed improvement is marginal, so it's not worth adding the complexity to the codebase.
image
